networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BeaconBlocksByRangeMessageHandler.java

```diff
@@ -92,17 +97,37 @@ public void onIncomingMessage(
   }
 
   private SafeFuture<RequestState> sendNextBlock(final RequestState requestState) {
-    return combinedChainDataClient
-        .getBlockAtSlotExact(requestState.currentSlot, requestState.headBlockRoot)
-        .thenCompose(
-            maybeBlock -> {
-              maybeBlock.ifPresent(requestState::sendBlock);
-              if (requestState.isComplete()) {
-                return completedFuture(requestState);
-              }
-              requestState.incrementCurrentSlot();
-              return sendNextBlock(requestState);
-            });
+    SafeFuture<Optional<SignedBeaconBlock>> blockFuture = loadNextBlock(requestState);
+    // Avoid risk of
```

Comment:
nit: completing the comment here would be nice its a pretty complex piece of code

Reply:
Fixed.